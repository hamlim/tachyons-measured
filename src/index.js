import { compose as ramdaCompose } from 'ramda';

export * from '../src/with-spacing';
export * from '../src/with-background-color';
export * from '../src/with-color';
export * from '../src/with-size';
export * from '../src/with-border';
export * from '../src/with-typography';
export * from '../src/with-defaults';
export * from '../src/with-base-styles';
export * from '../src/with-msrd';
export const compose = ramdaCompose;
