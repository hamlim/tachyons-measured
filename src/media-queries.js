import R from 'ramda';
import { PropTypes } from 'react';

export const supportedSelectors = ['ba', 'bt', 'br', 'bb', 'bl', 'bn', 'r',
  'rounded', 'bw', 'fw', 'h', 'lh', 'w', 'pa', 'pl', 'pr', 'pb', 'pt', 'pv',
  'ph', 'ma', 'ml', 'mr', 'mb', 'mt', 'mv', 'mh', 'na', 'nl', 'nr', 'nb', 'nt',
  'f'];

const mqProp = PropTypes.shape({
  all: PropTypes.any,
  ns: PropTypes.any,
  m: PropTypes.any,
  l: PropTypes.any,
});

export const addMQSupport = propType => PropTypes.oneOfType([
  propType,
  mqProp,
]);

const mqType = type => R.compose(R.equals(type), R.head);
const mqSelector = selector => ([type, val]) => `${selector}${val}-${type}`;

const invalidMQ = selector => () => {
  throw new Error(`Invalid media query type supplied to '${selector}'.
  It only supports all, ns, m & l.`);
};

export const mQObjToSelectors = R.curry((baseSelector, value) =>
  R.compose(
    R.map(R.cond([
      [mqType('all'), ([, val]) => `${baseSelector}${val}`],
      [mqType('ns'), mqSelector(baseSelector)],
      [mqType('m'), mqSelector(baseSelector)],
      [mqType('l'), mqSelector(baseSelector)],
      [R.T, invalidMQ(baseSelector)],
    ])),
    R.toPairs,
  )(value));
