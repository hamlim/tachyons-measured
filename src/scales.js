export const lineHeights = ['solid', 'title', 'copy'];

export const typeScale = [1, 2, 3, 4, 5, 6];

export const whiteSpaceScale = [0, 1, 2, 3, 4, 5, 6, '-headline',
  '-subheadline'];
export const negativeMarginScale = [1, 2, 3, 4, 5, 6];

export const widths = [1, 2, 3, 4, 5, '-10', '-20', '-25', '-30', '-33', '-34',
  '-40', '-50', '-60', '-70', '-75', '-80', '-90', '-100', '-third',
  '-two-thirds', '-auto'];

export const heights = [1, 2, 3, 4, 5, '-25', '-50', '-75', '-100', '-auto'];

export const border = {
  radii: [0, 1, 2, 3, 4, '-100', '-pill'],
  styles: ['none', 'dotted', 'dashed', 'solid'],
  widths: [0, 1, 2, 3, 4, 5],
};
