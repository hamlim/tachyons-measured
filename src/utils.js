import React from 'react';
import R from 'ramda';

export const isPresent = R.complement(R.isNil);
const hasPrototype = R.has('prototype');
const isReactComponent = R.propIs(Object, 'isReactComponent');
const isClassComponent = R.allPass([isPresent, hasPrototype, isReactComponent]);
const isNotClassComponent = R.complement(isClassComponent);
const doesNotHave = R.complement(R.has);

export const mapIfObj = fn => R.ifElse(R.is(Object),
  R.map(fn),
  fn,
);

/**
 * Compose classnames
 */
export const cx = R.compose(R.join(' '), R.flatten);

/**
 * Eagerly excute a component if it is a
 * referentially transparent function component
 *
 * Based on the createEagerElementUtil in Andrew Clark's recompose library
 * github.com/acdlite/recompose/blob/master/src/packages/recompose/utils
 */
const isReferentiallyTransparentFunctionComponent = R.allPass([
  R.is(Function),
  isNotClassComponent,
  doesNotHave('defaultProps'),
  doesNotHave('contextTypes'),
]);

export const createEagerElement = R.cond([
  [isReferentiallyTransparentFunctionComponent, R.call],
  [R.T, (Component, props) => <Component {...props} />],
]);

// construct the display name for a HOC
export const displayName = (component, name) =>
  `${name}(${component.displayName || component.name || component})`;

/**
 * Create a withStyle HOC by
 * providing a transformation
 */
export function createWithStyleHoc({
  name,
  transformation,
  propTypes,
  component,
}) {
  function WithStyle(ownerProps) {
    const props = transformation(ownerProps);
    return createEagerElement(component, props);
  }

  WithStyle.displayName = displayName(component, name);
  WithStyle.propTypes = R.merge(component.propTypes, propTypes);

  return WithStyle;
}
