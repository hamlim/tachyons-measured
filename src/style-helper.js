import R from 'ramda';
import { mQObjToSelectors } from './media-queries';

const isPresent = R.complement(R.isNil);
const hasMediaQuery = R.compose(R.is(Object), R.nth(1));

export const selectorFor = R.curry((prop, val) => `${prop}${val}`);

export const selectorForWithMQ = R.ifElse(hasMediaQuery,
  R.apply(mQObjToSelectors),
  R.apply(selectorFor),
);

/**
 * Get multiple tachyon classes
 * using an object where key is
 * selector type and value is value
 */
export const classesFor = R.compose(
  R.map(selectorForWithMQ),
  R.toPairs,
  R.filter(isPresent),
);

/**
 * classnames style compose
 * provide a map where key is classname
 * and value is a boolean
 */
export const composeClasses = R.compose(
  R.keys,
  R.filter(R.identity),
);
