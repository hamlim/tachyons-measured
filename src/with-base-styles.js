import R from 'ramda';
import PropTypes from 'prop-types';
import { cx, createWithStyleHoc } from './utils';

const sizePropTypes = {
  className: PropTypes.any,
};

const styleTransform = baseStyles => ({ className, ...ownerProps }) =>
  R.merge(
    { className: cx([baseStyles, className]) },
    ownerProps,
  );

export const withBaseStyles = styles => component => createWithStyleHoc({
  name: 'withBaseStyles',
  transformation: styleTransform(styles),
  propTypes: sizePropTypes,
  component,
});
