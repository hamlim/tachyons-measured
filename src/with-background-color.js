import R from 'ramda';
import { PropTypes } from 'react';
import { selectorFor } from './style-helper';
import { cx, createWithStyleHoc } from './utils';

const backgroundColorProps = colors => ({
  bg: PropTypes.oneOf(colors),
  className: PropTypes.any,
});

function backgroundColorTransform({ className, bg, ...ownerProps }) {
  return R.merge(
    { className: cx([selectorFor('bg-', bg), className]) },
    ownerProps,
  );
}

/**
 * HOC that allows you to configure
 * background color through props
 */
export const withBackgroundColor = colors => component => createWithStyleHoc({
  name: 'withWhiteSpace',
  transformation: backgroundColorTransform,
  propTypes: backgroundColorProps(colors),
  component,
});
