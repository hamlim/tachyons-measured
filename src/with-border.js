import R, { __ } from 'ramda';
import PropTypes from 'prop-types';
import { classesFor } from './style-helper';
import { cx, createWithStyleHoc, isPresent, normalizeClassNames } from './utils';
import { addMQSupport } from './media-queries';

/**
 * Scale
 */
export const border = {
  radii: [0, 1, 2, 3, 4, 100, 'pill'],
  styles: ['none', 'dotted', 'dashed', 'solid'],
  widths: [0, 1, 2, 3, 4, 5],
};

/**
 * Normalize classnames to add a `-`
 * prefix when needed
 */
const normalizeRadii = normalizeClassNames([100, 'pill']);

/**
 * Parse border declaration to get
 * border type and color values
 */
const hasMultipleDeclarations = R.compose(R.gt(__, 1), R.length);
const headOr = def => R.compose(R.defaultTo(def), R.head);
const warnAboutMultipleDeclarations = R.tap(() => {
  console.warn(`Warning: multiple border types declared.
    You can only declare one border type (ba, bl, br, bt, bb, bn) at a time.`);
});

const borderDeclaration = R.compose(
  headOr([]),
  R.ifElse(hasMultipleDeclarations,
    warnAboutMultipleDeclarations,
    R.identity,
  ),
  R.toPairs,
  R.filter(isPresent),
);

/**
 * withBorder HOC
 */
function borderTransform({
  className,
  ba, bl, br, bt, bb, bn,
  bw,
  radius,
  rounded,
  ...ownerProps
}) {
  const [type, value] = borderDeclaration({ ba, bl, br, bt, bb, bn });
  const borderColor = R.is(Boolean, value) ? undefined : value;

  const base = {
    br: normalizeRadii(radius),
    'br--': rounded,
    'b--': borderColor,
    bw,
  };

  const classList = type ? R.merge(base, { [type]: '' }) : base;

  const borderStyles = classesFor(classList);

  return R.merge(
    { className: cx([borderStyles, className]) },
    ownerProps,
  );
}

const borderType = colors => PropTypes.oneOfType([
  PropTypes.oneOf(colors),
  PropTypes.bool,
]);

const borderPropTypes = colors => ({
  ba: borderType(colors),
  bl: borderType(colors),
  br: borderType(colors),
  bt: borderType(colors),
  bb: borderType(colors),
  bn: PropTypes.bool,
  bw: addMQSupport(PropTypes.oneOf(border.widths)),
  radius: addMQSupport(PropTypes.oneOf(border.radii)),
  rounded: addMQSupport(PropTypes.oneOf(['bottom', 'top', 'right', 'left'])),
  className: PropTypes.any,
});

export const withBorder = colors => component => createWithStyleHoc({
  name: 'withBorder',
  transformation: borderTransform,
  propTypes: borderPropTypes(colors),
  component,
});
