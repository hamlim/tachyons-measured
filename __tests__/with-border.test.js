import React from 'react';
import renderer from 'react-test-renderer';
import { withBorder } from '../src/with-border';
import clrs from '../example/clrs';

const Button = withBorder(clrs)(props => <button {...props} />);
const Div = withBorder(clrs)('div');

test('Button', () => {
  const tree = renderer.create(
    <Button
      bb="gray" bw={{ l: 1, m: 2, ns: 3, all: 4 }}
      radius="pill"
      rounded="top"
      type="button"
      onClick={() => {}}
      className="myClass my-other-class"
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Div', () => {
  const tree = renderer.create(
    <Div
      ba bw={2}
      radius={{ l: 1, m: 2, ns: 100, all: 4 }}
      rounded={{ l: 'bottom', m: 'top', ns: 'right', all: 'left' }}
      className="myClass my-other-class"
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
