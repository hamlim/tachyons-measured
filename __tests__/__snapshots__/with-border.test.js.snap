// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button 1`] = `
<button
  className="br-pill br--top b--gray bw1-l bw2-m bw3-ns bw4 bb myClass my-other-class"
  onClick={[Function]}
  type="button"
/>
`;

exports[`Div 1`] = `
<div
  className="br1-l br2-m br-100-ns br4 br--bottom-l br--top-m br--right-ns br--left bw2 ba myClass my-other-class"
/>
`;
